Docs: Update contributor workflow to yarn rwt link (#1803)
redwoodjs/redwood: CONTRIBUTING.md

Here we'll call this codebase a "Redwood App"&mdash;it’s the fullstack-to-Jamstack solution you already know and love.

As a contributor, you'll have to gain familiarity with one more codebase: the Redwood Framework.
The Redwood Framework lives in the monorepo redwoodjs/redwood (which is where you're probably reading this); it contains all the packages that make Redwood Apps work the way they do.
After you use `git clone` to make a local copy, this is where you'll be making most of your changes.

> You'll probably want to see your changes to your local copy of the Redwood Framework “running live" in one of your own Redwood Apps or in one of our example apps.

Reviewer: ```suggestion
You'll probably want to see your changes to your local copy of the Redwood Framework "running live" in one of your own Redwood Apps or in one of our example apps.
```

> **I've used `yarn rw` before, but what's `yarn rwt`?**
>
> All workflows use `redwood-tools` (`rwt`), Redwood's companion CLI development tool.

All right, let's go through the entire contributor workflow, from Redwood Framework to Redwood App.

### Installing Dependencies

First, in your local copy of the Redwood Framework, make sure you've got the `main` branch's latest changes (via a `git pull`), then run `yarn install` in the root directory to install all the dependencies.

```terminal
cd redwood
yarn install
```

Then, in your Redwood App, which you've either just created using `yarn create redwood-app` or already had lying around, you'll want to upgrade to the canary&mdash;Redwood's bleeding edge&mdash;just so you can be sure you're testing your contribution with all the most recent changes:

```terminal
cd redwood-app # wherever your redwood-app happens to be, whether it's one of our templates or your own
yarn rw upgrade -t canary
```

### yarn rwt link

Ok, now that everything's up to date, in your Redwood App, run:

```bash
yarn rwt link [RW_PATH]
```

where [RW_PATH] is the path to your local copy of the Redwood Framework. For example,

```

> yarn rwt link ~/redwood

Reviewer: ```suggestion
yarn rwt link ../path/to/redwood
```
Just to be consistent with other example rw paths in this document

```

> You can set an env var so that you don't have to specify this. See [Specifying a `RW_PATH`](#specifying-a-rw_path).

You'll be prompted to add the redwood workspace to your Redwood App; go ahead and say yes.

> Then you'll start seeing a ton of output&mdash;`yarn rwt link` is building the Redwood Framework, watching it for changes, and copying all that over into the redwood workspace in your Redwood App. Your Redwood App isn't using the packages in `node_modules/@redwoodjs` anymore, but the packages in this workspace, with your local changes. You can even install packakges or upgrade dependencies&mdash;it's really that simple.

Reviewer: ```suggestion
Then you'll start seeing a ton of output&mdash;`yarn rwt link` is building the Redwood Framework, watching it for changes, and copying all that over into the redwood workspace in your Redwood App. Your Redwood App isn't using the packages in `node_modules/@redwoodjs` anymore, but the packages in this workspace, with your local changes. You can even install packages or upgrade dependencies&mdash;it's really that simple.
```

When you're done, make sure to `unlink` your Redwood App:

```
yarn rwt unlink
```

> This'll restore your Redwood App to it's original state, reverting it to the version of Redwood that it orignally had installed. Next time you want to contribute, just run `yarn rwt link` again!

Reviewer: "This will"  flows better off my tongue, but I guess it's personal preference.

Author: I opt to over-contract because it's closer to speech but I'm not rigid about it or anything if you think we should back off from it here--can make that change locally and push it up.

Reviewer: Your call. You're the native here 🙂   But I did just notice something that's most definitely wrong in the paragraph. It should be "its", not "it's" 😄

> This'll restore your Redwood App to it's original state, reverting it to the version of Redwood that it orignally had installed. Next time you want to contribute, just run `yarn rwt link` again!

Reviewer: ```suggestion
This'll restore your Redwood App to its original state, reverting it to the version of Redwood that it originally had installed. Next time you want to contribute, just run `yarn rwt link` again!
```

The Redwood Framework lives in the monorepo redwoodjs/redwood; it contains all the packages that make Redwood Apps work the way they do.

While you'll be making most of your changes in the Redwood Framework, you'll probably want to see your changes “running live" in one of your own Redwood Apps or in one of our example apps.

> Since we're always looking for ways to make contributing to Redwood as easy as possible, there's a few workflows for making this possible, but the one you'll want to use `yarn rwt link`. You can fall back on the others if that one doesn't work, but it'll make contributing a breeze.

Reviewer: ```suggestion
Since we're always looking for ways to make contributing to Redwood as easy as possible, there's a few workflows for making this possible, but the one you'll want to use is `yarn rwt link`. You can fall back on the others if that one doesn't work, but it'll make contributing a breeze.
```

> Since we're always looking for ways to make contributing to Redwood as easy as possible, there's a few workflows for making this possible, but the one you'll want to use `yarn rwt link`. You can fall back on the others if that one doesn't work, but it'll make contributing a breeze.

Reviewer: ```suggestion
Since we're always looking for ways to make contributing to Redwood as easy as possible, there's a few workflows for making this possible, but the one you'll want to use `yarn rwt link`. You can fall back on the others if that one doesn't work, but once you've tried `yarn rwt link` you won't want to.
```

Reading the original wording it was a bit unclear what "it'll" referred back to. It almost made it sound like falling back to something else would make contributing a breeze.

Author: Accidentally resolved this one; I'll make the change!

> **I've used `yarn rw` before but what's `yarn rwt`?**
>
> All workflows use `redwood-tools` (`rwt`), Redwood's companion CLI development tool.

Let's go through the entire contributor workflow, from Redwood Framework to Redwood App.

### Installing Dependencies

> First, in your local copy of the Redwood Framework, make sure you're on  `main` (via a `git pull`), then run `yarn install` in the root directory to install all the dependencies.

Reviewer: ```suggestion
First, in your local copy of the Redwood Framework, make sure you're on `main` (via a `git pull`), then run `yarn install` in the root directory to install all the dependencies.
```

```terminal
cd redwood
yarn install
```

Then, in your Redwood App, which you've either just created using `yarn create redwood-app` or already had lying around somewhere, you'll want to upgrade to the canary&mdash;Redwood's bleeding edge&mdash;just so you can be sure you're testing your contribution with all the most recent changes:

```terminal
cd redwood-app # wherever your redwood-app happens to be, whether it's one of our templates or your own
yarn rw upgrade -t canary
```

### yarn rwt link

Ok, now that everything's up to date, in your Redwood App, run:

```bash
yarn rwt link [RW_PATH]
```

where [RW_PATH] is the path to your local copy of the Redwood Framework. For exmaple,

```
yarn rwt link ~/redwood
```

You'll be prompted to add the redwood workspace to your Redwood App; go ahead and say yes.

> Then you'll start seeing a ton of output&mdash;`yarn rwt link` is building the Redwood Framework, watching it for changes, and copying all that over into a `redwood` workspace in your Redwood App.

Reviewer: ```suggestion
Then you'll start seeing a ton of output&mdash;`yarn rwt link` is building the Redwood Framework, watching it for changes, and copying all that over into the redwood workspace in your Redwood App.
```

It was "the redwood workspace" in the paragraph above, so I think it should be the same here

Your Redwood App isn't using the packages in `node_modules/@redwoodjs` anymore, but the packages in this workspace, with your minute to minute changes.

To see for yoruself, make a change in your local copy of the Redwood Framework and you'll see it live, propagated to your Redwood App. You can even install packakges. It really is that simple.

Make sure at the end to

```
yarn rwt restore
```

> This'll restore your Redwood App to it's original mundane state, stripping it of it's Redwood Framework powers. It's ok, it can enjoy those powers again when your contribution makes it into a release!

Reviewer 2: ```suggestion
This'll restore your Redwood App to it's original mundane state, stripping it of your local Redwood project - going back to the installed version of Redwood you orignally had in your project. Next time you want to contribute, just run `yarn rwt lin`k again!
```
